File: Anna Park/Week3/2025-06-22-Valid Parentheses.java
import java.util.*;

class Solution {
    public boolean isValid(String s) {
        
        Stack<Character> stack = new Stack<>();
        Map<Character, Character> brackets = new HashMap<>();

        brackets.put(']', '[');
        brackets.put('}', '{');
        brackets.put(')', '(');
        
        for (char c : s.toCharArray()){
            if (stack.isEmpty()){
                stack.push(c);
            } else{
                if (stack.peek() == brackets.get(c)){
                    stack.pop();
                }
            }
        }

        if (stack.isEmpty()){
            return true;
        } else{
            return false;
        }
    }
}

File: 2025/jimmycho/week3/2025-06-22-Valid-Parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        pairs = {
            "}" : "{",
            ")" : "(",
            "]" : "["
        }
        stack = []
        for char in s:
            if char in pairs.values():
                stack.append(char)
            elif not stack or pairs[char] != stack.pop():
                return False
        return False if stack else True